server/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -709,37 +709,46 @@ public void executeQueryPhase(
      * It is the responsibility of the caller to ensure that the ref count is correctly decremented
      * when the object is no longer needed.
      */
+    @SuppressWarnings({ "unchecked" })
     public void executeQueryPhase(QuerySearchRequest request, SearchShardTask task, ActionListener<QuerySearchResult> listener) {
         final ReaderContext readerContext = findReaderContext(request.contextId(), request.shardSearchRequest());
         final ShardSearchRequest shardSearchRequest = readerContext.getShardSearchRequest(request.shardSearchRequest());
         final Releasable markAsUsed = readerContext.markAsUsed(getKeepAlive(shardSearchRequest));
-        runAsync(getExecutor(readerContext.indexShard()), () -> {
-            readerContext.setAggregatedDfs(request.dfs());
-            try (
-                SearchContext searchContext = createContext(readerContext, shardSearchRequest, task, ResultsType.QUERY, true);
-                SearchOperationListenerExecutor executor = new SearchOperationListenerExecutor(searchContext)
-            ) {
-                searchContext.searcher().setAggregatedDfs(request.dfs());
-                QueryPhase.execute(searchContext);
-                if (searchContext.queryResult().hasSearchContext() == false && readerContext.singleSession()) {
-                    // no hits, we can release the context since there will be no fetch phase
-                    freeReaderContext(readerContext.id());
-                }
-                executor.success();
-                // Pass the rescoreDocIds to the queryResult to send them the coordinating node and receive them back in the fetch phase.
-                // We also pass the rescoreDocIds to the LegacyReaderContext in case the search state needs to stay in the data node.
-                final RescoreDocIds rescoreDocIds = searchContext.rescoreDocIds();
-                searchContext.queryResult().setRescoreDocIds(rescoreDocIds);
-                readerContext.setRescoreDocIds(rescoreDocIds);
-                searchContext.queryResult().incRef();
-                return searchContext.queryResult();
-            } catch (Exception e) {
-                assert TransportActions.isShardNotAvailableException(e) == false : new AssertionError(e);
-                logger.trace("Query phase failed", e);
-                // we handle the failure in the failure listener below
-                throw e;
-            }
-        }, wrapFailureListener(listener, readerContext, markAsUsed));
+        Rewriteable.rewriteAndFetch(
+            shardSearchRequest.getRewriteable(),
+            indicesService.getDataRewriteContext(shardSearchRequest::nowInMillis),
```

Comment:
Arg, I knew it was too easy.

We were not rewriting and fetching on one method and then unnecessarily creating scorers on two others:

1. `executeQueryPhase` when executing via the DFSQuery faze. This needs rewritten and rescorer fetched
2. `dfsPhase` needs rewritten and rescorers built, this is because we respect any queries that the rescorers have built to gather terms on them
3. `executeFetchPhase` for scrolls don't need rescorers built (cannot rescore and scroll)
4. `executeFetchPhase` for non-scrolls don't need rescorers built as it never runs them (its only fetching fields).

Reply:
@javanna && @dnhatn 

This adds an additional rewriteFetch phase ontop of the DFS query ID search.

The caller of this method is this:

```
transportService.registerRequestHandler(
            QUERY_ID_ACTION_NAME,
            ThreadPool.Names.SAME,
            QuerySearchRequest::new,
            (request, channel, task) -> {
                searchService.executeQueryPhase(request, (SearchShardTask) task, new ChannelActionListener<>(channel));
            }
        );
        TransportActionProxy.registerProxyAction(transportService, QUERY_ID_ACTION_NAME, true, QuerySearchResult::new);
```
Which is only used by DFS from what I can tell. Since the other query requests do this rewriteAndFetch from the calling thread, I did the same here and then switched to `runAsync` once we started executing on the shard (so we do the actual search in the Search thread, throttled search, etc.)

Is this OK? This is probably the trickiest part of this entire PR.

Reply:
Have you considered re-using the existing rewriteAndFetchShardRequest method in SearchService, but refactoring it to take in the context?

Reply:
> Have you considered re-using the existing rewriteAndFetchShardRequest method in SearchService, but refactoring it to take in the context?

There are additional checks present in that method.

```
if (request.readerId() != null) {
                l.onResponse(request);
            } else {
                shard.ensureShardSearchActive(b -> l.onResponse(request));
            }
```
That I didn't want to add here. It seemed safer to have a single duplication than add this check or additional predicates to handle which method is calling this when.

Reply:
Quick question: are we sure that we don't need the ensureShardSearchActive call? I would also have a preference for reusing the existing method, but I see why you went this way. I would reuse the method only if it can be fully reused, and I am wondering if that would be a good idea or can cause problems.


I will summarize my understanding of the context of this part of the change: we currently call rewriteAndFetch on the shard prior to executing dfs (dfs_query_then_fetch) or prior to executing the actual query (query_then_fetch). We are now adding an additional rewrite and fetch between right before executing the query but after the dfs phase has been completed. Does this sound accurate?

We need this because we never want to serialize models over the wire hence we want to retrieve them each time on the shard. Does this mean we will retrieve the same model on the same shard twice, once before dfs and once before the query execution?

Would it make sense to extract this specific change from this PR and open a separate one? I worry that despite a feature flag is mentioned, these are core changes to search execution that will affect the system regardless f whether they are used only behind a feature flag, at least potentially.

Reply:
> We are now adding an additional rewrite and fetch between right before executing the query but after the dfs phase has been completed. Does this sound accurate?

Correct.

> Does this mean we will retrieve the same model on the same shard twice, once before dfs and once before the query execution?

Correct. It should be in cache and thus only loaded from disk once. One optimization I have considered is somehow indicating via the DFS phase that we actually don't need the rescorer to be rewritten because we don't actually call rescore. But that started complicating things and I didn't want to introduce much more complexity in this PR than absolutely necessary. 

I think this would be a nice optimization in the future. But, since DFS and Query are ran on the same shards, it does seem like DFS is simply "pre-fetching" the model for the future query phase ;)

> Would it make sense to extract this specific change from this PR and open a separate one? 

I suppose I can do that. The main issue there is that the feature just won't work at all with DFS for the time being. I can mute the appropriate tests and open a new PR after this one adding this rewrite back.

> are we sure that we don't need the ensureShardSearchActive call?

I am not sure. But to me it seems like it should be in a separate PR with a separate discussion. We can always change this to reuse the existing method later.

Reply:
OK, I just pushed a reversion here. Not doing the rewrite in DFSQuery and muted the integration test for now (since it fails). Will open a new PR as soon as this is closed to continue this discussion there.

Reply:
@javanna && @jdconrad take a gander here: https://github.com/elastic/elasticsearch/pull/97152